DOCSP-15337 crud update (#73)
mongodb/docs-java: source/fundamentals/crud/write-operations/change-a-document.txt

=================
Change a Document
=================

.. default-domain:: mongodb

You can change documents in a MongoDB collection using two distinct
operation types: :ref:`update <update-operation>` and :ref:`replace

> <replace-operation>`. Update operations change specified fields in one

Reviewer: I think it's potentially confusing to the user to surface links without explaining what they are. It can be helpful to provide that explanation, but it might not fit so well in your current opening sentence. Consider rewording the opening sentence to identify what the user can learn in this guide/page and identifying them as sections. E.g. 
```
In this guide, you can learn how to change documents in a MongoDB collection using write operation types explained in the following sections:
- :ref:`Update <update-operation>`
...
```

or more documents using the
:java-docs:`updateOne() </apidocs/mongodb-driver-sync/com/mongodb/client/MongoCollection.html#updateOne(org.bson.conversions.Bson,org.bson.conversions.Bson)>` 
or :java-docs:`updateMany() </apidocs/mongodb-driver-sync/com/mongodb/client/MongoCollection.html#updateMany(org.bson.conversions.Bson,org.bson.conversions.Bson)>` 
method. A replace operation substitutes one document with a new
specified document using the
:java-docs:`replaceOne() </apidocs/mongodb-driver-sync/com/mongodb/client/MongoCollection.html#replaceOne(org.bson.conversions.Bson,TDocument)>` 
method.
    
In the following examples, a paint store sells eight different

> colors of paint. The store had their annual online sale. Their

Reviewer: I'm not sure this detail about their annual online sale makes sense here. This is the first time we're showing the data. It's also not clear whether any inventory was sold in the annual online sale. Instead, consider just mentioning that the collection represents their current inventory in the ``paint_inventory`` collection.

.. code-block:: json   

    AcknowledgedUpdateResult{ matchedCount=8, modifiedCount=8, upsertedId=null } 

The following shows the documents in the ``paint_inventory`` collection:

.. code-block:: json 

    { "_id": {"$oid": "606dfe52d506b97c576b22bc"}, "color": "red", "qty": 25 }
    { "_id": {"$oid": "606dfe52d506b97c576b22bd"}, "color": "purple", "qty": 28 }
    { "_id": {"$oid": "606dfe52d506b97c576b22be"}, "color": "blue", "qty": 20 }
    { "_id": {"$oid": "606dfe52d506b97c576b22bf"}, "color": "white", "qty": 20 }
    { "_id": {"$oid": "606dfe52d506b97c576b22c0"}, "color": "yellow", "qty": 26 }
    { "_id": {"$oid": "606dfe52d506b97c576b22c1"}, "color": "pink", "qty": 20 }
    { "_id": {"$oid": "606dfe52d506b97c576b22c2"}, "color": "green", "qty": 20 }
    { "_id": {"$oid": "606dfe52d506b97c576b22c3"}, "color": "black", "qty": 28 }

If zero documents matched from the update operation, there is no change
to the collection. You can configure an
:doc:`upsert </fundamentals/crud/write-operations/upsert>` to attempt an
update to a document or to insert a document if there are no matches. 

.. note::

    You cannot modify the ``_id`` field of a document nor change a field

>     to a value that violates a unique index constraint. See the MongoDB

Author: The the field it applies to is whichever the user specifies as the unique index (by default, it's ``_id``). I'll clarify it to say it cannot modify the unique index.

    server manual for more information on :manual:`unique indexes
    </core/index-unique/>`.

.. _replace-operation:

Replace
-------

To replace a document, create a new ``Document`` with the fields and
values that you would like to insert in your collection. This new

> document substitutes the document matched in your query filter. 

Reviewer: Currently, it's difficult to determine what's needed to perform the replace operation from this paragraph alone. Would it make more sense to provide the information in this order instead:

1) Reiterate what the replace operation does since this is the section on the replace method.
2) Mention the required parameters (query filter and replacement document)
3) Explain how to specify/instantiate a query filter
4) Explain how to specify/instantiate a replacement document

A new ``Document`` uses the following format: 

.. code-block:: java

>     Document document = new Document("<field>", "<value>");

Reviewer: We should be able to omit this if you mention what objects represent the query filter and replacement document. Instead, maybe you could replace this with the syntax of the replace operation or omit it if you think it's better in the example. Generally, I'd expect to see the syntax before an example.

Example
~~~~~~~

The store realizes they need to update their inventory again.  What they
thought was 20 cans of pink paint is actually 25 cans of orange paint. 

To update the inventory, query the ``paint_inventory`` collection where
the ``color`` is ``"pink"``, create a new ``Document`` where the ``color``

> is ``"orange"`` and ``qty`` is ``25``:

Reviewer: The presentation of this sentence sounds like the reader needs to perform sequential actions, but I think you're talking about specifying values for the parameters for a single operation. Could you make this more clear? It might help to format as a bulleted list of conditions. E.g. 
```
To update the inventory, call the replaceOne() method specifying the following: 
- A query filter that matches ...
- A replacement document that contains...
```

.. literalinclude:: /includes/fundamentals/code-snippets/Update.java
   :language: java
   :dedent:
   :start-after: begin replaceOneExample
   :end-before: end replaceOneExample

The method returns:

.. code-block:: json   

    AcknowledgedUpdateResult{ matchedCount=1, modifiedCount=1, upsertedId=null }

The following shows the updated document:

.. code-block:: json

    { "_id": { "$oid": "606dfe52d506b97c576b22c1" }, "color": "orange", "qty": 25 }

> If zero documents matched from the replace operation, there is
> no change to the collection. You can configure an 

Reviewer: Grammar fixes:
```suggestion
If zero documents matched the query filter in the replace operation, no changes are made to the collection. You can configure an 
```

:doc:`upsert </fundamentals/crud/write-operations/upsert>` to attempt an

> update to a document or to insert a document if there are no matches. 

Reviewer: I think upsert sounds unrelated to the rest of the paragraph. Maybe to address this you can:

1) create a new paragraph 
2) explain the behavior of an upsert
3) tell the reader that they can read the upsert guide for information on how to specify it

> If there are multiple documents matched for the ``replaceOne()`` method,
> the method will update the first occurrence of a match in the collection.

Reviewer: A few grammar fixes:

```suggestion
If multiple documents match the ``replaceOne()`` method,
the method replaces the first matched document matched by the query.
```

=================
Change a Document
=================

.. default-domain:: mongodb

In this page, you can learn how to change documents in a MongoDB collection
using two distinct operation types: 

- :ref:`Update <update-operation>`
- :ref:`Replace <replace-operation>`

Update operations change the value of the specified field in one or

> more documents using the 

Author: I think the mention of methods should go under each section opposed to stating it in the intro and the section.  I'll rework the intro to be simple statements about each operation and have a deeper explanation of them in their sections.

:java-docs:`updateOne() <apidocs/mongodb-driver-sync/com/mongodb/client/MongoCollection.html#updateOne(org.bson.conversions.Bson,org.bson.conversions.Bson)>` 
or :java-docs:`updateMany() <apidocs/mongodb-driver-sync/com/mongodb/client/MongoCollection.html#updateMany(org.bson.conversions.Bson,org.bson.conversions.Bson)>` 

> method. A replace operation substitutes one document with a new

Author: I'll be taking this part out of the intro, but will be including some of this in the replace section.

Update
------

Update operations specify an update document to change in one or more
documents that match your query filer using the 
:java-docs:`updateOne() <apidocs/mongodb-driver-sync/com/mongodb/client/MongoCollection.html#updateOne(org.bson.conversions.Bson,org.bson.conversions.Bson)>` 
or :java-docs:`updateMany() <apidocs/mongodb-driver-sync/com/mongodb/client/MongoCollection.html#updateMany(org.bson.conversions.Bson,org.bson.conversions.Bson)>` 
method. 

Update Operation Parameters:
~~~~~~~~~~~~~~~~~~~~~~~~~~~~

- Query filter: specifies the document you want to update in your collection
- Update document: an :doc:`Updates builder </fundamentals/builders/updates>` that specifies the field and value to change in the document matched

The update document uses the following format:

.. code-block:: java

    Bson update = Updates.<update operator>("<field>", "<value>");

See the MongoDB API documentation for a :java-core-api:`complete list of
update operators and their usage <com/mongodb/client/model/Updates.html>`.

The ``updateOne()`` and ``updateMany()`` method uses the following format: 

.. code-block:: java

    collection.updateOne(<query filter>, <update document>);

>     collection.updateMany(<query filter>, <update document>);

Author: I think including a return type the way you're stating may cause confusion since I mention in the intro that each method returns an AcknowledgementResult

Example
~~~~~~~

The paint store received a fresh shipment and needs to update their inventory.
The shipment contains 20 cans of each paint color.

To update the inventory, call the ``updateMany()`` method specifying the
following:

- A query filter that matches all the colors in the ``paint_inventory`` collection
- An update document that specifies to increment the ``qty`` field by "20"

.. literalinclude:: /includes/fundamentals/code-snippets/Update.java
   :language: java
   :dedent:
   :start-after: begin updateManyExample
   :end-before: end updateManyExample

The method returns:

.. code-block:: json   

>     AcknowledgedUpdateResult{ matchedCount=5, modifiedCount=5, upsertedId=null } 

Author: I addressed this by adding what the method returns in the intro. I'll provide a sentence following this to reiterate what it means.

.. code-block:: java

    collection.replaceOne(<query filter>, <new Document>);

Example
~~~~~~~

The paint store realizes they need to update their inventory again.  What they
thought was 20 cans of pink paint is actually 25 cans of orange paint. 

To update the inventory, call the ``replaceOne()`` method specifying the
following:

- A query filter that matches the ``color`` is "pink" in the ``paint_inventory`` collection
- A replacement document that contains: ``color`` is "orange" and ``qty`` is "25"

.. literalinclude:: /includes/fundamentals/code-snippets/Update.java
   :language: java
   :dedent:
   :start-after: begin replaceOneExample
   :end-before: end replaceOneExample

The method returns:

.. code-block:: json   

    AcknowledgedUpdateResult{ matchedCount=1, modifiedCount=1, upsertedId=null }

The following shows the updated document:

.. code-block:: json

    { "_id": 5, "color": "orange", "qty": 25 }

If zero documents matched the query filter in the replace operation,
there are no changes to the collection. See our :doc:`upsert
page</fundamentals/crud/write-operations/upsert>` to learn how you can
change the collection with zero documents matched. 

> If multiple documents match the ``replaceOne()`` method, the method
> replaces the first matched document matched by the query. 

Reviewer: I think the subject, verb, and objects of this sentence could be clarified by reordering the sentence..

```suggestion
If multiple documents match the query filter specified in the``replaceOne()`` method, it replaces the first result.
```

.. note::

>     You cannot modify the unique index of a document. See the MongoDB

Reviewer: In its current form, this note doesn't seem to contain enough information that helps the reader connect the operation with violating unique index constraints. Maybe a quick initial mention might help, e.g.
"Your replaceOne() operation cannot make changes to a document that violate unique index constraints on the collection."

query filter. 

The :java-docs:`updateOne() <apidocs/mongodb-driver-sync/com/mongodb/client/MongoCollection.html#updateOne(org.bson.conversions.Bson,org.bson.conversions.Bson)>` 
method changes the first document your query filter matches and the
:java-docs:`updateMany() <apidocs/mongodb-driver-sync/com/mongodb/client/MongoCollection.html#updateMany(org.bson.conversions.Bson,org.bson.conversions.Bson)>` 
method changes all the documents your query filter matches. 

You can call the ``updateOne()`` and ``updateMany()`` methods on a
``MongoCollection`` instance as follows: 

.. code-block:: java

    collection.updateOne(query, updateDocument);

    collection.updateMany(query, updateDocument);
    
Update Operation Parameters:
~~~~~~~~~~~~~~~~~~~~~~~~~~~~

- ``query`` specifies a query filter with the criteria to match documents to update in your collection
- ``updateDocument`` specifies the fields and values to change in the matching document or documents. For this example, we use the :doc:`Updates builder </fundamentals/builders/updates>` to create the update document.

You can create the ``updateDocument`` using an ``Updates`` builder as
follows:

.. code-block:: java

>     Bson updateDocument = Updates.<update operator>("<field>", "<value>");

Reviewer 2: It's confusing to show this as a Java code block, since `<SomeType>` is Java syntax for generics, so this use of `<..>` looks like a generic type instead of a placeholder for a method name.  Maybe just describe it instead?

Reviewer 2: Better, but still confusing because of <update operator>. I suggest just describing this in prose here, since we show an example right below, and there's a link to the Updates builders page as well.

Reviewer 2: I still think it's confusing because now it looks like `operator` is an actual method on `Updates`.  But I'll let it go.

    { "_id": 4, "color": "green", "qty": 26 }
    { "_id": 5, "color": "pink", "qty": 20 }

If zero documents match the query filter in the update operation,
``updateMany()`` makes no changes to documents in the collection. See
our :doc:`upsert guide </fundamentals/crud/write-operations/upsert>` to
learn how to insert a new document instead of updating one if no
documents match. 

.. note::

    The ``updateOne()`` and ``updateMany()`` methods cannot make changes
    to a document that violate unique index constraints on the
    collection. See the MongoDB server manual for more information on
    :manual:`unique indexes </core/index-unique/>`. 

.. _replace-operation:

Replace
-------

A replace operation substitutes one document from your collection. The
substitution occurs between a document your query filter matches and a
replacement document. 

The :java-docs:`replaceOne() <apidocs/mongodb-driver-sync/com/mongodb/client/MongoCollection.html#replaceOne(org.bson.conversions.Bson,TDocument)>` 
method removes all the existing fields and values in the
matching document (except the ``_id`` field) and substitutes it with the
replacement document. 

> You can call the ``repalceOne()`` method on a ``MongoCollection``

Reviewer 2: repalceOne => replaceOne

    { "_id": 3, "color": "yellow", "qty": 0 }
    { "_id": 4, "color": "green", "qty": 6 }
    { "_id": 5, "color": "pink", "qty": 0 }

.. _update-operation:

Update
------

Update operations can change the field and value. They apply changes
specified in an update document to one or more documents that match your
query filter. 

The :java-docs:`updateOne() <apidocs/mongodb-driver-sync/com/mongodb/client/MongoCollection.html#updateOne(org.bson.conversions.Bson,org.bson.conversions.Bson)>` 
method changes the first document your query filter matches and the
:java-docs:`updateMany() <apidocs/mongodb-driver-sync/com/mongodb/client/MongoCollection.html#updateMany(org.bson.conversions.Bson,org.bson.conversions.Bson)>` 
method changes all the documents your query filter matches. 

You can call the ``updateOne()`` and ``updateMany()`` methods on a
``MongoCollection`` instance as follows: 

.. code-block:: java

    collection.updateOne(query, updateDocument);

    collection.updateMany(query, updateDocument);
    
Update Operation Parameters:
~~~~~~~~~~~~~~~~~~~~~~~~~~~~

- ``query`` specifies the document you want to update in your collection

> - ``updateDocument`` specifies the field and value to change in the matching document or documents. For this example, we use the :doc:`Updates builder </fundamentals/builders/updates>` to create the update document.

Author: 1. To my understanding, whether it's updateOne or updateMany, only one field and value gets updated.  updateOne and updateMany only differ based on the number of documents that get updated.
2. I choose not to mention updateOptions since that's more specific to using upsert (and is mentioned on the upsert page) and it's not being used here

You can create the ``updateDocument`` using an ``Updates`` builder as
follows:

.. code-block:: java

    Bson updateDocument = Updates.<update operator>("<field>", "<value>");

See the MongoDB API documentation for a :java-core-api:`complete list of

> update operators and their usage <com/mongodb/client/model/Updates.html>`.

Reviewer: This link is to the Updates builder helper methods which may or may not correspond to the MongoDB-supported operators which you can find here https://docs.mongodb.com/manual/reference/operator/update/. Could you update the description or link to ensure this is accurate?

Example
~~~~~~~

The paint store receives a fresh shipment and needs to update their inventory.
The shipment contains 20 cans of each paint color.

To update the inventory, call the ``updateMany()`` method specifying the
following:

- A query filter that matches all the colors 
- An update document that contains instructions to increment the ``qty`` field by "20"

.. literalinclude:: /includes/fundamentals/code-snippets/Update.java
   :language: java
   :dedent:
   :start-after: begin updateManyExample
   :end-before: end updateManyExample

The method returns:

.. code-block:: json   

    AcknowledgedUpdateResult{ matchedCount=5, modifiedCount=5, upsertedId=null } 

The ``AcknowledgedUpdateResult`` tells us that five documents matched

> the query, five documents got modified and there was no upsert.

Reviewer: I think this sentence is passive voice. You can make it active voice by specifying what matched the query and what did the modifying.

The following shows the updated documents in the ``paint_inventory`` collection:

.. code-block:: json 

    { "_id": 1, "color": "red", "qty": 25 }
    { "_id": 2, "color": "purple", "qty": 28 }
    { "_id": 3, "color": "yellow", "qty": 20 }
    { "_id": 4, "color": "green", "qty": 26 }
    { "_id": 5, "color": "pink", "qty": 20 }

If zero documents match the query filter in the update operation,

> there are no changes to documents in the collection. See our :doc:`upsert

Reviewer: I think this sentence is passive voice. You can make it active voice by specifying what is not making changes to document in the collection.

Applies to the replaceOne() example description too.

guide </fundamentals/crud/write-operations/upsert>` to learn how to insert
a new document instead of updating one if no documents match. 

.. note::

    The ``updateOne()`` and ``updateMany()`` methods cannot make changes

>     to a document that  violate unique index constraints on the

Reviewer: ```suggestion
    to a document that violate unique index constraints on the
```
